yumietzk/eng-intern-assessment-react
# Shopify Technical Challenge - Frontend

## Description

This PR implements a stopwatch application satisfying the functionalities below and adds tests:

- The stopwatch should start counting when the user clicks the start button.
- The stopwatch should stop counting when the user clicks the stop button.
- The stopwatch should reset to zero when the user clicks the reset button.
- The stopwatch should record and display laps when the user clicks the lap button.

### The initial state
The time is rendered as "00:00:00.00." The Lap and Start buttons are rendered. The Lap button is disabled:
<img width="650" alt="Screenshot 2024-01-17 at 2 19 31 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/83822736-0dbf-4fc7-a976-8a8c90cc77b7">

### When the user clicks the Start button
The stopwatch starts counting. The Lap and Stop buttons are rendered:
<img width="650" alt="Screenshot 2024-01-17 at 2 20 14 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/6628ba53-649e-452b-a909-f5315a256bf5">

### When the user clicks the Stop button 
The stopwatch stops counting. The Reset and Start buttons are rendered:
<img width="650" alt="Screenshot 2024-01-17 at 2 24 29 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/3dd8bc2f-fd61-4aa0-beaf-1da9a369bbba">

### When the user clicks the Lap button
The laps are recorded and displayed. The laps can be scrolled as the number of laps increases:
<img width="650" alt="Screenshot 2024-01-17 at 2 26 54 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/624e1f9b-4053-4cd7-9959-884dae07b8b7">

### When the user clicks the Reset button
Same as the initial state. The time and laps are reset to zero. The Lap and Start buttons are rendered. The Lap button is disabled:
<img width="650" alt="Screenshot 2024-01-17 at 2 25 33 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/66d20721-de80-4171-a961-56c4a6f25d11">

### 🗂️ File structures

The file structures under the src folder are below:

- `App.tsx`: A main component that renders the stopwatch and handles its functionality
- `StopWatch.tsx`: A separate component that represents the stopwatch display
- `StopWatchButton.tsx`: A separate component that represents the start, stop, lap, and reset buttons
- `LapTable.tsx`: A separate component that represents the table containing all laps
- `LapRow.tsx`: A separate component that represents each lap
- `hooks/useTime.tsx`: A custom hook that represents logic
- `utils/formatTime.js`: A helper function that formats time
- `__test__`: A tests folder for components, a custom hook and a helper function

## Testing Steps

- Install all the dependencies: `npm install`
- Run the app locally: `npm start`. The app will run on [localhost:8080/](http://localhost:8080/). Test the app to make sure everything works as mentioned above in the description.

1. To test if the stopwatch starts counting: Click the Start button.
2. To test if the stopwatch stops counting: Click the Start and Stop buttons.
3. To test if the stopwatch records and displays laps: Click the Start button, then click the Lap button multiple times.
4. To test if the stopwatch resets everything: Click the Start button, then click the Lap button multiple times. And click the Stop and Reset buttons.

- Run tests for all files: `npm run test`. Check if all the test cases are passed as below:
<img width="941" alt="Screenshot 2024-01-17 at 3 23 26 PM" src="https://github.com/yumietzk/eng-intern-assessment-react/assets/61277579/9a7f0cf9-7f2c-4cca-9242-66b6bf1bc1e3">

* To run tests for a specific file: `npm run test -- <filename>`
